flink-state-backends/flink-statebackend-heap-spillable/src/test/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMapTest.java

```diff
@@ -115,577 +144,783 @@ public void testInitStateMap() {
+					kvIterator.remove();
 				}
 			}
 		}
+		return function.apply(removedCnt, removedStates);
+	}
 
-		// put some new states
-		for (long namespace = 10; namespace < 15; namespace++) {
-			for (int key = 0; key < 100; key++) {
-				totalSize++;
-				String state = String.valueOf(key * namespace);
-				if (random.nextBoolean()) {
-					stateMap.put(key, namespace, state);
-				} else {
-					assertNull(stateMap.putAndGetOld(key, namespace, state));
+	/**
+	 * Test state update operation.
+	 */
+	@Test
+	public void testUpdateState() {
+		testWithFunction(
+			(totalSize, stateMap, referenceStates) -> {
+				// update some states
+				for (Map.Entry<Long, Map<Integer, String>> namespaceEntry : referenceStates.entrySet()) {
+					long namespace = namespaceEntry.getKey();
+					for (Map.Entry<Integer, String> keyEntry : namespaceEntry.getValue().entrySet()) {
+						if (random.nextBoolean()) {
+							int key = keyEntry.getKey();
+							String state = keyEntry.getValue();
+							String newState = state + "-update";
+							keyEntry.setValue(newState);
+							if (random.nextBoolean()) {
+								stateMap.put(key, namespace, newState);
+							} else {
+								assertEquals(state, stateMap.putAndGetOld(key, namespace, newState));
+							}
+						}
+					}
 				}
-				referenceStates.computeIfAbsent(namespace, (none) -> new HashMap<>()).put(key, state);
-				assertEquals(totalSize, stateMap.size());
-				assertEquals(totalSize, stateMap.totalSize());
-			}
-		}
+				return getDefaultSizes(totalSize);
+			});
+	}
 
-		// remove some absent states
-		for (Map.Entry<Long, Set<Integer>> entry : removedStates.entrySet()) {
-			long namespace = entry.getKey();
-			for (int key : entry.getValue()) {
-				if (random.nextBoolean()) {
-					stateMap.remove(key, namespace);
-				} else {
-					assertNull(stateMap.removeAndGetOld(key, namespace));
+	/**
+	 * Test transform existing state.
+	 */
+	@Test
+	public void testTransformExistingState() throws Exception {
+		final AtomicReference<Exception> exceptionRef = new AtomicReference<>(null);
+		testWithFunction(
+			(totalSize, stateMap, referenceStates) -> {
+				StateTransformationFunction<String, Integer> function =
+					(String prevState, Integer value) -> prevState == null ? String.valueOf(value) : prevState + value;
+				// transform existing states
+				for (Map.Entry<Long, Map<Integer, String>> namespaceEntry : referenceStates.entrySet()) {
+					long namespace = namespaceEntry.getKey();
+					try {
+						for (Map.Entry<Integer, String> keyEntry : namespaceEntry.getValue().entrySet()) {
+							if (random.nextBoolean()) {
+								int key = keyEntry.getKey();
+								String state = keyEntry.getValue();
+								int delta = random.nextInt();
+								String newState = function.apply(state, delta);
+								keyEntry.setValue(newState);
+								stateMap.transform(key, namespace, delta, function);
+							}
+						}
+					} catch (Exception e) {
+						exceptionRef.set(e);
```

Comment:
would the exception be swallowed if we don't assign it to `execeptionRef`? If not, then I would prefer to let it simply bubble up.